platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/event/tipset/TipsetEventCreationManager.java

```diff
@@ -0,0 +1,242 @@
+     * @param startUpEventFrozenManager prevents event creation when the platform has just started up
+     */
+    public TipsetEventCreationManager(
+            @NonNull final PlatformContext platformContext,
+            @NonNull final ThreadManager threadManager,
+            @NonNull final Time time,
+            @NonNull final Random random,
+            @NonNull final Signer signer,
+            @NonNull final AddressBook addressBook,
+            @NonNull final NodeId selfId,
+            @NonNull final SoftwareVersion softwareVersion,
+            @NonNull final EventTransactionPool transactionPool,
+            @NonNull final Consumer<GossipEvent> newEventHandler,
+            @NonNull final QueueThread<EventIntakeTask> eventIntakeQueue,
+            @NonNull final Supplier<PlatformStatus> platformStatusSupplier,
+            @NonNull final StartUpEventFrozenManager startUpEventFrozenManager) {
+
+        this.newEventHandler = Objects.requireNonNull(newEventHandler);
+
+        Objects.requireNonNull(platformContext);
+        Objects.requireNonNull(threadManager);
+        Objects.requireNonNull(time);
+        Objects.requireNonNull(random);
+        Objects.requireNonNull(signer);
+        Objects.requireNonNull(addressBook);
+        Objects.requireNonNull(selfId);
+        Objects.requireNonNull(softwareVersion);
+        Objects.requireNonNull(transactionPool);
+        Objects.requireNonNull(eventIntakeQueue);
+        Objects.requireNonNull(platformStatusSupplier);
+        Objects.requireNonNull(startUpEventFrozenManager);
+
+        final TipsetEventCreator baseCreator = new TipsetEventCreatorImpl(
+                platformContext,
+                time,
+                random /* does not need to be cryptographically secure */,
+                signer,
+                addressBook,
+                selfId,
+                softwareVersion,
+                transactionPool);
+
+        eventCreator = new ThrottledTipsetEventCreator(
+                platformContext,
+                time,
+                transactionPool,
+                eventIntakeQueue,
+                platformStatusSupplier,
+                startUpEventFrozenManager,
+                baseCreator);
+
+        final EventCreationConfig eventCreationConfig =
+                platformContext.getConfiguration().getConfigData(EventCreationConfig.class);
+
+        workQueue = new MultiQueueThreadConfiguration(threadManager)
+                .setThreadName("event-creator")
+                .setCapacity(eventCreationConfig.creationQueueSize())
+                .setMaxBufferSize(eventCreationConfig.creationQueueBufferSize())
+                .addHandler(EventImpl.class, this::handleEvent)
+                .addHandler(Long.class, this::handleMinimumGenerationNonAncient)
+                .setIdleCallback(this::maybeCreateEvent)
+                .setBatchHandledCallback(this::maybeCreateEvent)
```

Comment:
If I understand correctly, this thread has two opportunities to attempt creating an event:

1. Every `eventCreationConfig.creationQueueWaitForWorkPeriod()` if there is no work
2. After it has handled a batch of work


If the queue is always full because this node is slow, or being spammed, it will create fewer events than it would otherwise. This is ok, because if it has a big backlog of events, any new event will have old parents and may not reach consensus. Should there be a check in the event creation logic to see how big the size of the queue is before creating an event?

Reply:
We discussed this offline, repeating discussion for others.

Even if this queue is getting spammed, I think we still want to create an event at least once per buffer it it is legal to do so according to the tipset algorithm. Otherwise, a node that is slightly slower than its peers could be forced into a perpetual state of not creating events.